docs/design/proxy-support.md

```diff
@@ -0,0 +1,201 @@
+# Supporting package managers proxies in Hermeto
+
+## Background
+
+It might be desirable to pull dependencies through a proxy for a variety of
+reasons (builds reproducibility, additional verification, fetch speedup,
+etc). This means that Hermeto has to support proxying packages for
+individual package managers.
+
+## Proxy usage overview
+
+Most supported package managers (with a notable exception of Yarn v1) make
+use of proxy URL and support standard mechanisms for authentication.  Proxy
+usage is transparent to users as long as they provide correct proxy
+address and authentication credentials. Given that a proxy is correctly
+configured a user would just receive their dependencies as usual with an
+underlying package manager handling everything seamlessly.  Thus a way to
+consume proxy URL and credentials from the environment must be added to
+Hermeto.
+
+A proxy might require authentication or not.  A proxy might exist for just
+some package managers and do not exist for others. This means that each
+individual package manager will need to handle checking the proxy setting
+and authentication setting. Once these settings are known they must be
+made available to the underlying tool used for fetching the actual
+dependencies either via environment or via a configuration option.
+
+
+## Additional considerations
+
+### SBOM enhancement
+
+SBOM must be marked to indicate whether a proxy was used or not:
+
+```
+   proxy: None | proxy_url
```

Comment:
A recent observation, we may need to make this a **list** of URLs if this is directly to translate to individual backends, see `GOPROXY` in https://go.dev/ref/mod#environment-variables, it can be a comma separated list of proxies and unfortunately for us you have pretty much no way of telling which one was used unless you sniff the network.

Reply:
Yes, this is the best we can provide in such case. In practice this feature is sometimes used to implement a poor man's retry: one would specify the same proxy several times in hopes that if there is a transient error during the first few attempts then it would autoresolve by the time the end of list is reached. I doubt that specifying multiple different proxies is a widespread practice. Even if this happens then those who would need to investigate a build would need to study several proxies instead of not knowing of their existence.